File: learn-api/app/models/task.py
from sqlalchemy import JSON, Boolean, Column, ForeignKey, Integer, Text
from sqlalchemy.orm import relationship

from app.db.base_class import Base


class Task(Base):
    id = Column(Integer, primary_key=True, index=True)
    layer = Column(Integer, nullable=False)
    base_task_id = Column(Integer, ForeignKey("basetask.id"))
    task_type = Column(Integer)
    annotation_type = Column(Integer, nullable=False)
    min_correct = Column(Integer, nullable=False, default=1)
    task_question = Column(Text)
    knowledge_level = Column(Integer)
    solution = Column(JSON, nullable=True)
    task_data = Column(JSON, nullable=True)
    info_annotations = Column(JSON, nullable=True)
    annotation_groups = Column(JSON, nullable=True)
    hints = relationship(
        "TaskHint", cascade="all, delete-orphan", order_by="TaskHint.needed_mistakes"
    )
    can_be_solved = Column(Boolean, nullable=True, default=True)
    questionnaires = relationship(
        "Questionnaire", secondary="taskquestionnaires", back_populates="tasks"
    )

File: frontend/src/model/task/task.ts
import { TaskHint } from './taskHint';
import { UserSolution } from '../userSolution';
import { AnnotationData } from '../viewer/export/annotationData';
import { AnnotationGroup } from './annotationGroup';
import { TaskType } from '../../core/types/taskType';
import { Questionnaire } from 'model/questionnaires/questionnaire';

export interface Task {
  id: number;
  layer: number;
  task_type: TaskType;
  task_question: string;
  knowledge_level: number;
  task_data?: AnnotationData[] | string[];
  user_solution?: UserSolution;
  solution?: AnnotationData[] | number[] | string[];
  info_annotations?: any[];
  annotation_type: number;
  min_correct: number;
  annotation_groups: AnnotationGroup[];
  hints: TaskHint[];
  can_be_solved: boolean;
  questionnaires: Questionnaire[];
}

export interface TaskCreate {
  layer: number;
  task_question: string;
  base_task_id: number;
  task_type?: number;
  solution?: any;
  task_data?: any;
  knowledge_level: number;
  annotation_type: number;
  min_correct: number;
  annotation_groups?: AnnotationGroup[];
  hints: TaskHint[];
  can_be_solved?: boolean;
  info_annotations?: any[];
  questionnaires: Questionnaire[];
}

export interface TaskUpdate {
  task_id: number;
  layer?: number;
  task_question?: string;
  min_correct?: number;
  knowledge_level?: number;
  task_data?: string;
  solution?: string;
  annotation_groups?: AnnotationGroup[];
  can_be_solved?: boolean;
  info_annotations?: any[];
}
